tools/async-job.py

```diff
@@ -13,47 +13,33 @@
 # This script handles both batch replication and batch creation of revisions,
 # depending on by which name it is called.
 
-NAME          = os.path.basename(sys.argv[0])
-LOCKFILE_PATH = '/tmp/irods-{}.lock'.format(NAME)
+# When created or modified data objects are given a random balance id between 1-64.
+# This script can be run to handle replications or revisions within a range that is passed to the script.
+# Making it possible to have multiple replication/revision processes running in parallel where each process covers its own range.
 
+NAME          = os.path.basename(sys.argv[0])
 
 def get_args():
     """Parse command line arguments"""
     parser = argparse.ArgumentParser(description='Yoda replication and revision job')
     parser.add_argument('--verbose', '-v', action='store_const', default="0", const="1",
                     help='Log more information in rodsLog for troubleshooting purposes')
+    parser.add_argument('--balance_id_min', type=int, default=1,
+                    help='Minimal balance id to be handled by this job (range 1-64) for load balancing purposes')
+    parser.add_argument('--balance_id_max', type=int, default=64,
+                    help='Maximum balance id to be handled by this job (range 1-64) for load balancing purposes')
     return parser.parse_args()
 
 
-def lock_or_die():
-    """Prevent running multiple instances of this job simultaneously"""
-
-    # Create a lockfile for this job type, abort if it exists.
-    try:
-        fd = os.open(LOCKFILE_PATH, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
-    except OSError:
-        if os.path.exists(LOCKFILE_PATH):
-            print('error: Lock file {} exists'.format(LOCKFILE_PATH), file=sys.stderr)
-            exit(1)
-        else:
-            raise
-    os.write(fd, str(os.getpid()))
-    os.close(fd)
-
-    # Remove lock no matter how we exit.
-    atexit.register(lambda: os.unlink(LOCKFILE_PATH))
-
-
 if 'replicate' in NAME:
-    rule_name = 'uuReplicateBatch(*verbose)'
+    rule_name = 'uuReplicateBatch(*verbose, *balance_id_min, *balance_id_max)'
 elif 'revision' in NAME:
-    rule_name = 'uuRevisionBatch(*verbose)'
+    rule_name = 'uuRevisionBatch(*verbose, *balance_id_min, *balance_id_max)'
 else:
     print('bad command "{}"'.format(NAME), file=sys.stderr)
     exit(1)
 
 args = get_args()
-lock_or_die()
```

Comment:
I believe we need to insert a replacement for the old locking function here, in order to reduce the risk of multiple revision/replication jobs running simultaneously with overlapping balance ID ranges.

Reply:
This is now implemented where the locking file is extended with the corresponding min/max balance_id. This, as this is directly deterministic for the job.